quinn-proto/src/endpoint.rs

```diff
@@ -586,14 +586,57 @@ impl Endpoint {
         let was_closed = self.connections[conn.0].state.as_ref().unwrap().is_closed();
 
         // State transitions
-        let state = self.connections[conn.0].state.take().unwrap();
-        let state = self.connections[conn.0].handle_connected_inner(
+        let prev_state = self.connections[conn.0].state.take().unwrap();
+        let was_handshake = match prev_state {
+            State::Handshake(_) => true,
+            _ => false,
+        };
+        let state = match self.connections[conn.0].handle_connected_inner(
             &mut self.ctx,
             now,
             remote,
             packet,
-            state,
-        );
+            prev_state,
+        ) {
+            Ok(state) => state,
+            Err(conn_err) => {
+                self.ctx.events.push_back((
+                    conn,
+                    Event::ConnectionLost {
+                        reason: conn_err.clone(),
+                    },
+                ));
+
+                match conn_err {
+                    ConnectionError::ApplicationClosed { reason } => {
+                        if was_handshake {
+                            State::handshake_failed(reason, None)
+                        } else {
+                            State::closed(reason)
+                        }
+                    }
+                    ConnectionError::ConnectionClosed { reason } => {
+                        if was_handshake {
+                            State::handshake_failed(reason, None)
+                        } else {
+                            State::closed(reason)
+                        }
+                    }
+                    ConnectionError::TransportError { error_code } => {
+                        if was_handshake {
+                            State::handshake_failed(error_code, None)
+                        } else {
+                            State::closed(error_code)
+                        }
+                    }
+                    ConnectionError::VersionMismatch => State::Draining,
+                    _ => {
+                        debug!(self.ctx.log, "unexpected connection error received"; "error" => %conn_err, "initial_conn_id" => %self.connections[conn.0].initial_id);
+                        return;
```

Comment:
This leaves the connection's state empty, which is assumed e.g. on linel 589 to be impossible and will likely lead to a panic down the line. We should probably either panic immediately, or invent an appropriate state for the remaining cases (reset and timeout, so draining, perhaps?). I don't have reason to believe we'll emit either of those from `handle_connection_inner` in the future, so I'd lean towards panic. Probably best to explicitly enumerate the remaining cases either way, since any new additions might need different handling.

Reply:
Yeah, good call.